### DISCUSSION TITLE
(spring extension) Converters not registered.

### DISCUSSION THREAD

**Discussion by basven** [mapstruct/mapstruct#3336] [Q&A] [answered]
I used mapstruct years ago and am picking it up for this existing project I joined.  I am trying something pretty basic and defined the following mappers:

```
@Mapper(config = MapperSpringConfig.class)
public interface TimeKeeperDtoMapper extends Converter<TimeKeeperDTO, TimeKeeper> {
    @Mapping(target = "budget", ignore = true)
    @Override
    TimeKeeper convert(@NonNull TimeKeeperDTO dto);
```
    
and another going the opposite direction:
    
```
@Mapper(config = MapperSpringConfig.class)
public interface TimeKeeperMapper extends Converter<TimeKeeper, TimeKeeperDTO> {
    @Mapping(target = "isPrimary", source = "primary")
    @Override
    @Nullable
    TimeKeeperDTO convert(@NonNull TimeKeeper entity);
}
```
    
where my config class looks as follows:
    
```
@MapperConfig(componentModel = MappingConstants.ComponentModel.SPRING, uses = LitigationConversionAdapter.class)
@SpringMapperConfig(conversionServiceAdapterClassName ="LitigationConversionAdapter", conversionServiceBeanName = "litigationConversionAdapter")
public interface MapperSpringConfig {
}
```
    
    
I started out with trying to inject the ConversionService into a bean but was getting a spring error that I had 2 conversion beans in my path: mvcConversionService and defaultConversionService.
I then added the Qualifier annotation to specify one of them.  The problem is that for either one of them, it seems my ConversionAdapter class was not registered since calling:
    
```
conversionService.convert(TimeKeeper.builder().name("test").build(), TimeKeeperDTO.class);
```
    
always results in this exception:
    
```
org.springframework.core.convert.ConverterNotFoundException: No converter found capable of converting from type [timeKeeper] to type [TimeKeeperDTO]
	at org.springframework.core.convert.support.GenericConversionService.handleConverterNotFound(GenericConversionService.java:322)
	at org.springframework.core.convert.support.GenericConversionService.convert(GenericConversionService.java:195)
	at org.springframework.core.convert.support.GenericConversionService.convert(GenericConversionService.java:175)
```
    
Do I need to define some Configuration class?  It must be something simple I am missing.

**Comment by filiphr** [comment]
@Chessray is this something that you might be able to help with?

**Comment by Chessray** [comment]
Could you tell us a bit more about the project setup? A standard Spring Boot project typically picks this up out of the box. @basven Just to clarify: From Spring's point of view, everything is registered as a Bean, but depending on your setup you might need to add the converters to the conversion service separately. See the tests for [examples](https://github.com/mapstruct/mapstruct-spring-extensions/blob/main/examples/noconfig/src/test/java/org/mapstruct/extensions/spring/example/ConversionServiceAdapterIntegrationTest.java) of that.

**Comment by basven** [comment]
This project uses spring boot 2.7.6 but I noticed it also pulled in a
dependency on spring cloud. I don't think it needs that so i will see what
happens when I remove that.  I will also look at those tests you
mentioned,  I missed that.

**Comment by Chessray** [comment, marked as answer]
Most Spring Boot projects will configure automatically. But AFAIR this requires the web module. If your project doesn't use that, I recommend looking at [this](https://github.com/mapstruct/mapstruct-spring-extensions/blob/main/examples/simple-springboot/src/main/java/org/mapstruct/extensions/spring/example/boot/config/ConverterConfig.java) example.

**Comment by basven** [comment]
Pretty sure we have the web starter dependency

**Comment by filiphr** [comment]
Thanks @Chessray for looking into this. @basven if it doesn't work like @Chessray explained then please provide us a minimal reproducible example that we can try out.

**Comment by Chessray** [comment]
One more thing that strikes me as _very_ unusual is that you mentioned having _two_ `ConversionService` beans in your application context. I can imagine this might confuse the whole setup. You may want to explore that angle as well. In any case, please let us know the outcome. This may result in some type of change to the extension.

**Reply by basven** [reply]
Apologies for the delayed response; slammed at work and in the middle of moving.  Anyhow, I created a small test project and as soon as I add the below dependency I am getting the NoUniqueBeanDefinition exception.  Removing it and it just works.
```   
<dependency>
      <groupId>org.springframework.boot</groupId>
      <artifactId>spring-boot-starter-data-rest</artifactId>
</dependency>
```

Note that when I remove the above dependency but add `@EnableWebMvc` to my application, I see the same exception again.

Moving forward, I will take a look at the examples on how to register the converters to the service.

**Reply by Chessray** [reply]
Thanks. So combining various default starters can actually result in multiple `ConversionService`s being in the `ApplicationContext`. I guess at the very least that would call for an option to specify a bean name at the injection point inside the adapter class. Of course, it still wouldn't solve the auto-registration of the individual `Converter`s. In any case, this has brought up something to work on for the extension project.

**Reply by Chessray** [reply]
I just realized that we at least provide the option of specifying the bean name already: https://mapstruct.org/documentation/spring-extensions/reference/html/#customConversionService

Since I never used this feature myself, I'm unsure as to whether there's an elegant way to register the individual `Converter`s with one of several `ConversionService`s automatically. If @dmham86 is still around: Are you able to give a detailed explanation?

**Comment by basven** [comment]
This worked for me.  Defined a bean in my main app like this:
```
    @Bean("mapStructConversionService")
    ConfigurableConversionService getConversionService(PersonMapper personMapper) {
        DefaultConversionService service = new DefaultConversionService();
        service.addConverter(personMapper);
        return service;
    }
```
Inject and test it:

```
@SpringBootTest
public class MapperTest {

    @Inject
    @Qualifier("mapStructConversionService")
    private ConversionService conversionService;

    @Test
    public void test(){
        Person p = new Person(1, "test");
        PersonDTO dto = conversionService.convert(p, PersonDTO.class);
        assertNotNull(dto);
        assertEquals(p.getId(), dto.getId());
        assertEquals(p.getName(), dto.getName());

    }
}
```
Of course the registration might become a nightmare when there are a lot of converters.

**Comment by basven** [comment]
The better solution is the one @Chessray suggested later, which is [this one](https://github.com/mapstruct/mapstruct-spring-extensions/blob/main/examples/simple-springboot/src/main/java/org/mapstruct/extensions/spring/example/boot/config/ConverterConfig.java)
Thanks for all the help.
